### ISSUE TITLE
Error in db.dart for Collection.streamData

### ISSUE THREAD

**Issue by FireJuun** [fireship-io/flutter-firebase-quizapp-course#11]
db.dart provides an excellent, abstract means to connect Firestore documents and collections to Flutter without all the boilerplate. However, I noticed a bug when working with streams in db.dart. When creating a stream collection, such as by calling:
`StreamProvider<List<Example>>.value(value: Global.exampleRef.streamData())`

...inside of a MultiProvider widget, you will get the following error message:

> An exception was thrown by _MapStrean<QuerySnapshot, List<Example>> listened by StreamProvider<List<Example>>, but no 'catchError was provided.
> 
> Exception: type 'MappedListIterable<DocumentSnapshot, Example>' is not a subtype of type 'List<Example>'

In db.dart, Line 47 is missing a .toList() at the end:
`Stream<List<T>> streamData() {
    return ref.snapshots().map((list) =>
        list.documents.map((doc) => Global.models[T](doc.data) as T).toList());
  }`

**Comment by maliii3** [comment]
Hey, I might have found the solution for your problem. I was facing the same problem as you did, while browsing in the depth of the StackOverflow, I found this [.](https://stackoverflow.com/a/49603938) Hope it helps after 11 months :)

**Comment by FireJuun** [comment]
Haha. Thanks. :) 

Adding .toList() to the end of line 47 was the answer, by the way. I'm not sure if I had made that clear in my description of the issue. Technically, I probably should have just submitted a PR with the solution/update...

**Comment by lucygliang** [comment]
Thank you! I was running into the same issue and adding the toList() worked for me.

**Comment by jeswinkjessor** [comment]
**Hi, I was trying to fetch data from firestore using provider and this exception was thrown, have no idea what went wrong.**

                ════════ Exception caught by provider ══════════
The following assertion was thrown:
An exception was throw by _MapStream<QuerySnapshot, List

listened by
StreamProvider<List>, but no catchError was provided.
Exception:
type 'int' is not a subtype of type 'double'

**ADDRESS CLASS**
Address.fromFirestore(Map<String, dynamic> firestore)
: addressID = firestore['addressID'],
streetNumber = firestore['streetNumber'],
streetName = firestore['streetName'],
city = firestore['city'],
province = firestore['province'],
customerContact = firestore['customerContact'],
numberOfBedrooms = firestore['numberOfBedrooms'],
numberOfFamilyRooms = firestore['numberOfFamilyRooms'],
squareFeet = firestore['squareFeet'],
customerName = firestore['customerName'],
appointmentDate = DateFormat("yyyy-MM-dd").parse(firestore['appointmentDate']),
nextAppointmentDate = DateFormat("yyyy-MM-dd").parse(firestore['nextAppointmentDate']),
price = firestore['price'];

**DATABASE SERVICE**
Stream<List

getAddress(){
return _db.collection('addressCollection').snapshots().map((snapshot) =>
snapshot.docs.map((document) => Address.fromFirestore(document.data())).toList());
}
**WIDGET CLASS**
@override
Widget build(BuildContext context) {
final address = Provider.of<List

>(context, listen: false);
}
**MAIN**
providers: [
ChangeNotifierProvider(create: (context) => AddressProvider(),),
StreamProvider(create: (context) => firestoreService.getAddress()),
],

****I have used another provider in the same project for different collection and it worked fine, just the address is giving hard time**

**Comment by jrheisler** [comment]
When you upgrade your firestore you'll wind up with a different set of errors with one related to the streamData. I can't figure it out.

  Stream<List<T>> streamData() {
    return ref.snapshots().map((list) => list.documents.map((doc) => Global.models[T](doc.data) as T) );
  }

gives:
Expected a value of type 'Map<dynamic, dynamic>', but got one of type '() => Map<String, dynamic>'

**Comment by FireJuun** [comment]
> When you upgrade your firestore you'll wind up with a different set of errors with one related to the streamData. I can't figure it out.
> 
> Stream<List> streamData() {
> return ref.snapshots().map((list) => list.documents.map((doc) => Global.models[T](doc.data) as T) );
> }
> 
> gives:
> Expected a value of type 'Map<dynamic, dynamic>', but got one of type '() => Map<String, dynamic>'

You need to add a () somewhere in your code. The error is that it wants a map as its return value, but it is instead receiving a function that creates a map... Hence the () before => Map...

Assuming everything worked before the upgrade, you may want to consider changing doc.data to doc.data()

@jrheisler let me know if that was your issue.

**Comment by jrheisler** [comment]
Exactly, but I changed it on the receiving end in the model to data(). Thank you!!!

**Comment by LongKhoa0706** [comment]
> > When you upgrade your firestore you'll wind up with a different set of errors with one related to the streamData. I can't figure it out.
> > Stream streamData() {
> > return ref.snapshots().map((list) => list.documents.map((doc) => Global.models[T](doc.data) as T) );
> > }
> > gives:
> > Expected a value of type 'Map<dynamic, dynamic>', but got one of type '() => Map<String, dynamic>'
> 
> You need to add a () somewhere in your code. The error is that it wants a map as its return value, but it is instead receiving a function that creates a map... Hence the () before => Map...
> 
> Assuming everything worked before the upgrade, you may want to consider changing doc.data to doc.data()
> 
> @jrheisler let me know if that was your issue.

it working <3!

**Comment by jrheisler** [comment]
@LongKhoa0706 that was exactly how I fixed the issue on my side.

**Comment by yussifm** [comment]
======== Exception caught by provider ==============================================================
The following assertion was thrown:
An exception was throw by _MapStream<QuerySnapshot<Map<String, dynamic>>, List<ProductModel>> listened by

StreamProvider<List<ProductModel>>, but no `catchError` was provided.

Exception:
Bad state: field does not exist within the DocumentSnapshotPlatform

====================================================================================================

**Comment by igbokwenu** [comment]
> ======== Exception caught by provider ============================================================== The following assertion was thrown: An exception was throw by _MapStream<QuerySnapshot<Map<String, dynamic>>, List> listened by
> 
> StreamProvider<List>, but no `catchError` was provided.
> 
> Exception: Bad state: field does not exist within the DocumentSnapshotPlatform
> 
> ====================================================================================================

I am having this exact error. Were you able to find a fix?
